File: apps/ios/PhrenLive/Sources/PhrenLive/GatewayConnections.swift
import Crypto
import Foundation
import NIOCore
import NIOPosix
import NIOSSH
import PhrenKit

/// One authenticated SSH connection per computer, shared by every gateway
/// request. A fresh connection costs a TCP handshake, a key exchange and an
/// authentication round trip; each request is then one session channel on
/// the shared connection, which is what makes a chat send feel immediate.
///
/// sshd allows ten sessions per connection, so once eight channels are open
/// (long-lived transcript and terminal streams count) a request gets a
/// dedicated connection instead of an open failure. A connection that closes
/// or goes idle leaves the pool; the next request reconnects.
final class GatewayConnections: @unchecked Sendable {
    static let shared = GatewayConnections()
    static let maximumChildren = 8
    static let idleSeconds: Int64 = 90

    /// Channel bookkeeping is confined to the connection's event loop.
    final class Connection: @unchecked Sendable {
        let loop: EventLoop
        let channel: Channel
        let ssh: NIOSSHHandler
        let pooled: Bool
        fileprivate(set) var children = 0
        fileprivate var idle: Scheduled<Void>?
        init(loop: EventLoop, channel: Channel, ssh: NIOSSHHandler, pooled: Bool) {
            self.loop = loop; self.channel = channel; self.ssh = ssh; self.pooled = pooled
        }
    }

    private let lock = NSLock()
    private var ready: [String: Connection] = [:]
    private var connecting: [String: Task<Connection, Error>] = [:]
    /// How many connections were opened; tests read it to prove reuse.
    private(set) var opened = 0

    static func key(host: LiveHost, key: Curve25519.Signing.PrivateKey) -> String {
        let device = Data(SHA256.hash(data: key.publicKey.rawRepresentation)).base64EncodedString()
        return [host.id.uuidString, host.address, String(host.port), host.username, host.fingerprint ?? "", device].joined(separator: "|")
    }

    /// An open connection with room for one more channel, or a new one. The
    /// caller owns one child slot on the result until it calls `release`.
    func connection(for host: LiveHost, key: Curve25519.Signing.PrivateKey) async throws -> Connection {
        let poolKey = Self.key(host: host, key: key)
        if let existing = pooled(poolKey), try await existing.reserve() { return existing }
        let task: Task<Connection, Error> = lock.withLock {
            if let inFlight = connecting[poolKey] { return inFlight }
            let task = Task { try await self.connect(host: host, key: key, pooled: true) }
            connecting[poolKey] = task
            return task
        }
        do {
            let connection = try await task.value
            lock.withLock { if connecting[poolKey] == task { connecting[poolKey] = nil } }
            if try await connection.reserve() { register(connection, poolKey: poolKey); return connection }
            // The shared connection filled up while it was being opened.
            register(connection, poolKey: poolKey)
        } catch {
            lock.withLock { if connecting[poolKey] == task { connecting[poolKey] = nil } }
            throw error
        }
        let dedicated = try await connect(host: host, key: key, pooled: false)
        _ = try await dedicated.reserve()
        return dedicated
    }

    /// Gives the child slot back. A dedicated connection closes with its
    /// only channel; a pooled one waits idle for the next request.
    func release(_ connection: Connection, healthy: Bool) {
        connection.loop.execute {
            connection.children = max(0, connection.children - 1)
            if !connection.pooled || !healthy {
                connection.channel.close(promise: nil)
                self.remove(connection)
                return
            }
            guard connection.children == 0 else { return }
            connection.idle?.cancel()
            connection.idle = connection.loop.scheduleTask(in: .seconds(Self.idleSeconds)) {
                guard connection.children == 0 else { return }
                connection.channel.close(promise: nil)
                self.remove(connection)
            }
        }
    }

    /// Drops everything: tests use it between servers.
    func reset() {
        let connections = lock.withLock { () -> [Connection] in
            let values = Array(ready.values); ready = [:]; connecting = [:]; return values
        }
        for connection in connections { connection.channel.close(promise: nil) }
    }

    private func pooled(_ poolKey: String) -> Connection? {
        lock.withLock { ready[poolKey] }
    }

    private func register(_ connection: Connection, poolKey: String) {
        lock.withLock {
            // A connection that died before registration is not kept.
            guard connection.channel.isActive else { return }
            ready[poolKey] = connection
        }
    }

    private func remove(_ connection: Connection) {
        lock.withLock {
            if let key = ready.first(where: { $0.value === connection })?.key { ready[key] = nil }
        }
    }

    private func connect(host: LiveHost, key: Curve25519.Signing.PrivateKey, pooled: Bool) async throws -> Connection {
        try host.validate()
        let loop = MultiThreadedEventLoopGroup.singleton.next()
        let ready = Exchange(result: loop.makePromise(of: Data.self))
        let deadline = loop.scheduleTask(in: .seconds(15)) { ready.finish(.failure(LiveConnectionError.timeout)) }
        ready.result.futureResult.whenComplete { _ in deadline.cancel() }
        let opened = ConnectionParts()
        let bootstrap = ClientBootstrap(group: loop).connectTimeout(.seconds(10)).channelInitializer { channel in
            channel.eventLoop.makeCompletedFuture {
                let ssh = NIOSSHHandler(
                    role: .client(.init(
                        userAuthDelegate: DeviceAuthentication(username: host.username, key: key, exchange: ready),
                        serverAuthDelegate: PinnedHost(fingerprint: host.fingerprint)
                    )), allocator: channel.allocator,
                    inboundChildChannelInitializer: { channel, _ in
                        channel.eventLoop.makeFailedFuture(LiveConnectionError.disconnected)
                    })
                try channel.pipeline.syncOperations.addHandlers(ssh, GatewayConnectionReady(exchange: ready) { [weak self] in
                    guard let self, let connection = opened.connection else { return }
                    self.remove(connection)
                })
                opened.channel = channel; opened.ssh = ssh
            }
        }
        bootstrap.connect(host: host.address, port: host.port).whenFailure { ready.finish(.failure($0)) }
        do {
            _ = try await ready.result.futureResult.get()
        } catch {
            opened.channel?.close(promise: nil)
            throw error
        }
        guard let channel = opened.channel, let ssh = opened.ssh, channel.isActive else { throw LiveConnectionError.disconnected }
        let connection = Connection(loop: loop, channel: channel, ssh: ssh, pooled: pooled)
        opened.connection = connection
        lock.withLock { self.opened += 1 }
        return connection
    }
}

/// Filled on the event loop during connection setup.
private final class ConnectionParts: @unchecked Sendable {
    var channel: Channel?
    var ssh: NIOSSHHandler?
    var connection: GatewayConnections.Connection?
}

extension GatewayConnections.Connection {
    /// Claims a child slot when the connection is alive and has room.
    fileprivate func reserve() async throws -> Bool {
        try await loop.submit {
            guard self.channel.isActive, self.children < GatewayConnections.maximumChildren else { return false }
            self.children += 1
            self.idle?.cancel(); self.idle = nil
            return true
        }.get()
    }
}

/// Reports the parent connection's authentication and closure.
private final class GatewayConnectionReady: ChannelInboundHandler {
    typealias InboundIn = Any
    let exchange: Exchange
    let onClose: () -> Void
    init(exchange: Exchange, onClose: @escaping () -> Void) { self.exchange = exchange; self.onClose = onClose }
    func userInboundEventTriggered(context: ChannelHandlerContext, event: Any) {
        if event is UserAuthSuccessEvent { exchange.finish(.success(Data())) }
        context.fireUserInboundEventTriggered(event)
    }
    func errorCaught(context: ChannelHandlerContext, error: Error) {
        exchange.finish(.failure(error))
        context.close(promise: nil)
    }
    func channelInactive(context: ChannelHandlerContext) {
        exchange.finish(.failure(LiveConnectionError.disconnected))
        onClose()
        context.fireChannelInactive()
    }
}

File: apps/android/phrenkit/src/main/kotlin/com/phren/kit/live/GatewayConnections.kt
package com.phren.kit.live

import com.phren.kit.LiveHost
import net.schmizz.sshj.DefaultConfig
import net.schmizz.sshj.SSHClient
import net.schmizz.sshj.common.Buffer
import net.schmizz.sshj.common.Ed25519KeyFactory
import net.schmizz.sshj.connection.channel.direct.Session
import net.schmizz.sshj.transport.TransportException
import net.schmizz.sshj.transport.verification.HostKeyVerifier
import net.schmizz.sshj.userauth.UserAuthException
import net.schmizz.sshj.userauth.keyprovider.KeyPairWrapper
import org.bouncycastle.crypto.params.Ed25519PrivateKeyParameters
import java.net.SocketTimeoutException
import java.security.MessageDigest
import java.security.PublicKey
import java.util.Base64
import java.util.concurrent.Executors
import java.util.concurrent.ScheduledFuture
import java.util.concurrent.TimeUnit

/** The device's Ed25519 identity for one computer (DeviceSSHKey.swift), from its 32-byte seed. */
class DeviceKey(val seed: ByteArray) {
    init { require(seed.size == 32) { "An Ed25519 seed is 32 bytes." } }
    val publicRaw: ByteArray = Ed25519PrivateKeyParameters(seed, 0).generatePublicKey().encoded

    /** `ssh-ed25519 <base64 blob>`: the OpenSSH public key line. */
    val openSSHPublicKey: String get() = "ssh-ed25519 " + Base64.getEncoder().encodeToString(sshBlob(publicRaw))

    /** The restricted line the computer's authorized_keys gets: only the dispatcher runs. */
    fun authorizedKey(comment: String = "phren-android") =
        "restrict,pty,command=\"sh ~/.local/share/phren/bridge/dispatch\" $openSSHPublicKey $comment"

    internal fun keyProvider() = KeyPairWrapper(Ed25519KeyFactory.getPublicKey(publicRaw), Ed25519KeyFactory.getPrivateKey(seed))

    companion object {
        fun generate(): DeviceKey = DeviceKey(ByteArray(32).also { java.security.SecureRandom().nextBytes(it) })

        private fun sshBlob(raw: ByteArray): ByteArray {
            val type = "ssh-ed25519".toByteArray()
            return java.nio.ByteBuffer.allocate(8 + type.size + raw.size).putInt(type.size).put(type).putInt(raw.size).put(raw).array()
        }

        /** "SHA256:" + unpadded base64 of the key blob, as `ssh-keygen -l` prints it (PhrenConnection.fingerprint). */
        fun fingerprint(blob: ByteArray): String =
            "SHA256:" + Base64.getEncoder().withoutPadding().encodeToString(MessageDigest.getInstance("SHA-256").digest(blob))

        fun fingerprint(publicKeyLine: String): String? {
            val fields = publicKeyLine.split(' ')
            if (fields.size < 2) return null
            return runCatching { fingerprint(Base64.getDecoder().decode(fields[1])) }.getOrNull()
        }

        fun fingerprint(key: PublicKey): String = fingerprint(Buffer.PlainBuffer().putPublicKey(key).compactData)
    }
}

/**
 * One authenticated SSH connection per computer, shared by every request
 * (GatewayConnections.swift). Each request is one session channel; sshd allows
 * ten per connection, so past eight open channels a request gets a dedicated
 * connection. An idle connection closes after 90 s; one a request found broken
 * is retired: it takes no new channels and closes when its last one finishes.
 */
object GatewayConnections {
    const val MAXIMUM_CHILDREN = 8
    const val IDLE_SECONDS = 90L
    const val DRAIN_SECONDS = 65L

    class Connection internal constructor(val client: SSHClient, val pooled: Boolean, internal val poolKey: String) {
        internal var children = 0
        internal var retired = false
        internal var idle: ScheduledFuture<*>? = null
        val isActive: Boolean get() = client.isConnected && client.isAuthenticated
    }

    private val lock = Any()
    private val ready = mutableMapOf<String, Connection>()
    private val connecting = mutableMapOf<String, Any>()
    private val timers = Executors.newSingleThreadScheduledExecutor { Thread(it, "phren-ssh-idle").apply { isDaemon = true } }
    /** How many connections were opened; tests read it to prove reuse. */
    @Volatile var opened = 0; private set

    fun poolKey(host: LiveHost, key: DeviceKey): String {
        val device = Base64.getEncoder().encodeToString(MessageDigest.getInstance("SHA-256").digest(key.publicRaw))
        return listOf(host.id.toString().uppercase(), host.address, host.port.toString(), host.username, host.fingerprint ?: "", device).joinToString("|")
    }

    /** An open connection with room for one more channel, or a new one. The caller owns one slot until [release]. Blocking. */
    fun connection(host: LiveHost, key: DeviceKey): Connection {
        val poolKey = poolKey(host, key)
        // One dial per computer at a time; the others wait for it and share it.
        val gate = synchronized(lock) { connecting.getOrPut(poolKey) { Any() } }
        synchronized(gate) {
            synchronized(lock) { ready[poolKey]?.let { if (reserve(it)) return it } }
            val fresh = connect(host, key, pooled = true, poolKey)
            synchronized(lock) {
                if (fresh.isActive) ready[poolKey] = fresh
                if (reserve(fresh)) return fresh
            }
        }
        // The shared connection filled up: a dedicated one for this request.
        val dedicated = connect(host, key, pooled = false, poolKey)
        synchronized(lock) { reserve(dedicated) }
        return dedicated
    }

    private fun reserve(connection: Connection): Boolean {
        if (!connection.isActive || connection.retired || connection.children >= MAXIMUM_CHILDREN) return false
        connection.children++
        connection.idle?.cancel(false); connection.idle = null
        return true
    }

    /** Gives the slot back. A dedicated connection closes with its only channel; a pooled one waits idle. */
    fun release(connection: Connection, healthy: Boolean) = synchronized(lock) {
        connection.children = maxOf(0, connection.children - 1)
        if (!healthy && !connection.retired) {
            connection.retired = true
            remove(connection)
            connection.idle?.cancel(false)
            connection.idle = timers.schedule({ close(connection) }, DRAIN_SECONDS, TimeUnit.SECONDS)
        }
        if (!connection.pooled || connection.retired) {
            if (connection.children == 0) { connection.idle?.cancel(false); connection.idle = null; remove(connection); close(connection) }
            return@synchronized
        }
        if (connection.children != 0) return@synchronized
        connection.idle?.cancel(false)
        connection.idle = timers.schedule({
            synchronized(lock) { if (connection.children == 0) { remove(connection); close(connection) } }
        }, IDLE_SECONDS, TimeUnit.SECONDS)
    }

    /** Drops everything: the app's reconnect and tests. */
    fun reset() {
        val all = synchronized(lock) { ready.values.toList().also { ready.clear() } }
        all.forEach(::close)
    }

    private fun remove(connection: Connection) { if (ready[connection.poolKey] === connection) ready.remove(connection.poolKey) }
    private fun close(connection: Connection) { runCatching { connection.client.disconnect() } }

    /** Records what the host offered, so a refusal can say whether it was unknown or changed. */
    private class PinnedHost(val fingerprint: String?) : HostKeyVerifier {
        @Volatile var failure: LiveConnectionError? = null
        override fun verify(hostname: String?, port: Int, key: PublicKey): Boolean {
            val received = runCatching { DeviceKey.fingerprint(key) }.getOrNull()
            failure = when {
                received == null -> LiveConnectionError.ChangedHost()
                fingerprint == null -> LiveConnectionError.UntrustedHost(received)
                fingerprint != received -> LiveConnectionError.ChangedHost()
                else -> null
            }
            return failure == null
        }
        override fun findExistingAlgorithms(hostname: String?, port: Int): List<String> = emptyList()
    }

    private val config by lazy { DefaultConfig() }

    private fun connect(host: LiveHost, key: DeviceKey, pooled: Boolean, poolKey: String): Connection {
        host.validate()
        val client = SSHClient(config)
        val verifier = PinnedHost(host.fingerprint)
        client.addHostKeyVerifier(verifier)
        client.connectTimeout = 10_000
        client.connection.keepAlive.keepAliveInterval = 30
        try {
            client.connect(host.address, host.port)
            client.auth(host.username, net.schmizz.sshj.userauth.method.AuthPublickey(key.keyProvider()))
        } catch (error: Exception) {
            runCatching { client.disconnect() }
            throw verifier.failure ?: when (error) {
                is UserAuthException -> LiveConnectionError.Authentication()
                is SocketTimeoutException -> LiveConnectionError.Timeout()
                is TransportException -> LiveConnectionError.Ssh("The SSH exchange failed (${error.message}). Reconnect to the computer and try again.")
                is java.io.IOException -> LiveConnectionError.Ssh("Could not reach the computer (${error.message}). Check Tailscale and SSH.")
                else -> error
            }
        }
        synchronized(lock) { opened++ }
        return Connection(client, pooled, poolKey)
    }

    /** The restricted helper command; everything else is the Hook's HTTP. */
    internal fun openPipe(connection: Connection): Pair<Session, Session.Command> {
        val session = try { connection.client.startSession() } catch (error: Exception) {
            throw LiveConnectionError.Ssh("The computer refused the SSH channel (${error.message}). Check Phren Hook and available SSH sessions.")
        }
        return try { session to session.exec("phren-hook v1 pipe") } catch (error: Exception) {
            runCatching { session.close() }
            throw LiveConnectionError.Disconnected()
        }
    }
}
